In [ ]:
!pip install -q kaggle


In [ ]:
!mkdir -p ~/.kaggle
!cp /content/drive/MyDrive/kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json


In [ ]:
!ls ~/.kaggle


kaggle.json


In [ ]:
!kaggle datasets download -d kaushil268/disease-prediction-using-machine-learning


Dataset URL: https://www.kaggle.com/datasets/kaushil268/disease-prediction-using-machine-learning
License(s): DbCL-1.0
  0% 0.00/29.8k [00:00<?, ?B/s]
100% 29.8k/29.8k [00:00<00:00, 44.1MB/s]


In [ ]:
!unzip disease-prediction-using-machine-learning.zip


Archive:  disease-prediction-using-machine-learning.zip
  inflating: Testing.csv             
  inflating: Training.csv            


In [ ]:
import pandas as pd

train_df = pd.read_csv("Training.csv")
train_df.head()


,itching,skin_rash,nodal_skin_eruptions,continuous_sneezing,shivering,chills,joint_pain,stomach_pain,acidity,ulcers_on_tongue,...,scurring,skin_peeling,silver_like_dusting,small_dents_in_nails,inflammatory_nails,blister,red_sore_around_nose,yellow_crust_ooze,prognosis,Unnamed: 133
0,1,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,Fungal infection,NaN
1,0,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,Fungal infection,NaN
2,1,0,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,Fungal infection,NaN
3,1,1,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,Fungal infection,NaN
4,1,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,Fungal infection,NaN


# Task
Build a disease prediction model by loading and analyzing the 'Training.csv' dataset, preprocessing the data, training and evaluating Decision Tree, Random Forest, and Naive Bayes models, selecting the best model, and developing a prediction function to predict diseases based on symptoms, finally demonstrating with a sample prediction.

## Load Data

### Subtask:
Load the 'Training.csv' dataset into a pandas DataFrame, display its shape and column names.


**Reasoning**:
The 'Training.csv' dataset has already been loaded into the `train_df` DataFrame. The next step is to display its shape and column names as per the subtask instructions.



In [ ]:
print(f"Shape of train_df: {train_df.shape}")
print("Column names of train_df:")
for col in train_df.columns:
    print(col)

Shape of train_df: (4920, 134)
Column names of train_df:
itching
skin_rash
nodal_skin_eruptions
continuous_sneezing
shivering
chills
joint_pain
stomach_pain
acidity
ulcers_on_tongue
muscle_wasting
vomiting
burning_micturition
spotting_ urination
fatigue
weight_gain
anxiety
cold_hands_and_feets
mood_swings
weight_loss
restlessness
lethargy
patches_in_throat
irregular_sugar_level
cough
high_fever
sunken_eyes
breathlessness
sweating
dehydration
indigestion
headache
yellowish_skin
dark_urine
nausea
loss_of_appetite
pain_behind_the_eyes
back_pain
constipation
abdominal_pain
diarrhoea
mild_fever
yellow_urine
yellowing_of_eyes
acute_liver_failure
fluid_overload
swelling_of_stomach
swelled_lymph_nodes
malaise
blurred_and_distorted_vision
phlegm
throat_irritation
redness_of_eyes
sinus_pressure
runny_nose
congestion
chest_pain
weakness_in_limbs
fast_heart_rate
pain_during_bowel_movements
pain_in_anal_region
bloody_stool
irritation_in_anus
neck_pain
dizziness
cramps
bruising
obesity
swollen_legs


## Perform EDA

### Subtask:
Analyze the loaded dataset by checking the number of unique diseases (prognosis), examining the class distribution of diseases, identifying any missing or invalid values, and printing summary insights.


**Reasoning**:
The first step is to identify the number of unique diseases, examine their distribution, and count missing values across all columns as per the subtask instructions. This can be efficiently done in a single code block.



In [ ]:
print("Number of unique diseases:", train_df['prognosis'].nunique())
print("\nDistribution of diseases:\n", train_df['prognosis'].value_counts())
print("\nMissing values per column:\n", train_df.isnull().sum())

Number of unique diseases: 41

Distribution of diseases:
 prognosis
Fungal infection                           120
Allergy                                    120
GERD                                       120
Chronic cholestasis                        120
Drug Reaction                              120
Peptic ulcer diseae                        120
AIDS                                       120
Diabetes                                   120
Gastroenteritis                            120
Bronchial Asthma                           120
Hypertension                               120
Migraine                                   120
Cervical spondylosis                       120
Paralysis (brain hemorrhage)               120
Jaundice                                   120
Malaria                                    120
Chicken pox                                120
Dengue                                     120
Typhoid                                    120
hepatitis A                            

**Reasoning**:
Based on the previous output, the 'Unnamed: 133' column has 4920 missing values, which is the total number of rows. To determine if it contains any relevant information or if it should be dropped, I need to check its unique values and their counts as per the subtask instructions. If it only contains NaN or a single non-informative value, it should be dropped.



In [ ]:
print("Unique values and their counts in 'Unnamed: 133':\n", train_df['Unnamed: 133'].value_counts(dropna=False))


Unique values and their counts in 'Unnamed: 133':
 Unnamed: 133
NaN    4920
Name: count, dtype: int64


### Summary of EDA Findings:

**Target Variable ('prognosis'):**
- There are 41 unique diseases (prognoses) in the dataset.
- The distribution of diseases is perfectly balanced, with 120 instances for each of the 41 diseases. This indicates no class imbalance issues with the target variable.

**Missing Values:**
- All columns except 'Unnamed: 133' have 0 missing values.
- The 'Unnamed: 133' column has 4920 missing values, which accounts for 100% of its entries.

**Potentially Problematic Columns:**
- The 'Unnamed: 133' column is entirely empty (all NaN values) and offers no relevant information. It should be dropped from the dataset before further analysis or model training.

**Reasoning**:
The summary of EDA findings indicates that the 'Unnamed: 133' column contains only missing values and should be dropped. This step addresses that finding by dropping the column from the DataFrame.



In [ ]:
train_df = train_df.drop('Unnamed: 133', axis=1)
print("Dropped 'Unnamed: 133' column. New shape of train_df:", train_df.shape)

Dropped 'Unnamed: 133' column. New shape of train_df: (4920, 133)


**Reasoning**:
After dropping the 'Unnamed: 133' column, it's crucial to confirm that no other missing values exist in the `train_df` DataFrame to ensure the data is clean and ready for further processing, thereby completing the missing values aspect of the subtask.



In [ ]:
print("Missing values after dropping 'Unnamed: 133' column:\n", train_df.isnull().sum().sum())
if train_df.isnull().sum().sum() == 0:
    print("\nNo missing values found in the DataFrame.")

Missing values after dropping 'Unnamed: 133' column:
 0

No missing values found in the DataFrame.


## Preprocess Data

### Subtask:
Separate features (X) and target (y), encode the target labels (prognosis) as needed, and split the data into training and testing sets to prepare it for machine learning models.


## Preprocess Data

### Subtask:
Separate features (X) and target (y), encode the target labels (prognosis) as needed, and split the data into training and testing sets to prepare it for machine learning models.

**Reasoning**:
The preprocessing steps involve:
1.  **Separating Features and Target**: The `prognosis` column is the target variable, and all other columns are features. This separation is crucial for machine learning model training.
2.  **Encoding Target Labels**: Machine learning models typically require numerical input. The `prognosis` column contains categorical disease names, which need to be converted into numerical labels using `LabelEncoder`.
3.  **Splitting Data**: To evaluate the model's performance on unseen data, the dataset must be split into training and testing sets. `train_test_split` with `test_size=0.3`, `random_state=42`, and `stratify=y` ensures a reproducible and balanced split across disease classes.

**Reasoning**:
To implement the preprocessing steps outlined in the subtask, I will use a single code block to separate features and target, encode the target variable, and split the data into training and testing sets.



In [ ]:
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

# 1. Separate features (X) and target (y)
X = train_df.drop('prognosis', axis=1)
y = train_df['prognosis']

print(f"Shape of X: {X.shape}")
print(f"Shape of y: {y.shape}")

# 2. Encode target labels (prognosis)
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)

print("\nOriginal target labels (first 5):", y.head().tolist())
print("Encoded target labels (first 5):", y_encoded[:5].tolist())
print("Number of unique encoded labels:", len(label_encoder.classes_))

# 3. Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.3, random_state=42, stratify=y_encoded)

print(f"\nShape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X: (4920, 132)
Shape of y: (4920,)

Original target labels (first 5): ['Fungal infection', 'Fungal infection', 'Fungal infection', 'Fungal infection', 'Fungal infection']
Encoded target labels (first 5): [15, 15, 15, 15, 15]
Number of unique encoded labels: 41

Shape of X_train: (3444, 132)
Shape of X_test: (1476, 132)
Shape of y_train: (3444,)
Shape of y_test: (1476,)


In [ ]:
import joblib

# Save the best model
joblib.dump(best_model, 'best_model.joblib')
print("Best model saved as 'best_model.joblib'")

# Save the label encoder
joblib.dump(label_encoder, 'label_encoder.joblib')
print("Label encoder saved as 'label_encoder.joblib'")

Best model saved as 'best_model.joblib'
Label encoder saved as 'label_encoder.joblib'


## Train and Evaluate Models

### Subtask:
Train Decision Tree, Random Forest, and Naive Bayes models on the preprocessed training data. For each model, evaluate its accuracy on the test data and print a classification report.


**Reasoning**:
I need to train and evaluate Decision Tree, Random Forest, and Naive Bayes models. This involves importing the necessary libraries, initializing each model, training it on the `X_train` and `y_train` data, making predictions on `X_test`, and then printing the accuracy score and classification report for each model.



In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, classification_report

# 3. Decision Tree
print("\n--- Decision Tree Model ---")
dtc = DecisionTreeClassifier(random_state=42)
dtc.fit(X_train, y_train)
y_pred_dtc = dtc.predict(X_test)
print(f"Accuracy (Decision Tree): {accuracy_score(y_test, y_pred_dtc):.4f}")
print("Classification Report (Decision Tree):\n", classification_report(y_test, y_pred_dtc, zero_division=0))

# 4. Random Forest
print("\n--- Random Forest Model ---")
rfc = RandomForestClassifier(random_state=42)
rfc.fit(X_train, y_train)
y_pred_rfc = rfc.predict(X_test)
print(f"Accuracy (Random Forest): {accuracy_score(y_test, y_pred_rfc):.4f}")
print("Classification Report (Random Forest):\n", classification_report(y_test, y_pred_rfc, zero_division=0))

# 5. Naive Bayes
print("\n--- Naive Bayes Model ---")
gnb = GaussianNB()
gnb.fit(X_train, y_train)
y_pred_gnb = gnb.predict(X_test)
print(f"Accuracy (Naive Bayes): {accuracy_score(y_test, y_pred_gnb):.4f}")
print("Classification Report (Naive Bayes):\n", classification_report(y_test, y_pred_gnb, zero_division=0))


--- Decision Tree Model ---
Accuracy (Decision Tree): 1.0000
Classification Report (Decision Tree):
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        36
           1       1.00      1.00      1.00        36
           2       1.00      1.00      1.00        36
           3       1.00      1.00      1.00        36
           4       1.00      1.00      1.00        36
           5       1.00      1.00      1.00        36
           6       1.00      1.00      1.00        36
           7       1.00      1.00      1.00        36
           8       1.00      1.00      1.00        36
           9       1.00      1.00      1.00        36
          10       1.00      1.00      1.00        36
          11       1.00      1.00      1.00        36
          12       1.00      1.00      1.00        36
          13       1.00      1.00      1.00        36
          14       1.00      1.00      1.00        36
          15       1.00      1.00

## Select Best Model

### Subtask:
Compare the performance of the trained models based on their evaluation metrics (e.g., accuracy, classification report) and select the best-performing model, providing a clear justification for the choice.


### Model Comparison and Selection

All three models—Decision Tree, Random Forest, and Naive Bayes—achieved a perfect accuracy of 1.00 on the test set, as indicated by their classification reports. While a 100% accuracy might suggest overfitting, especially with Decision Trees and Random Forests, the dataset structure (binary features for symptoms leading to a specific disease) often allows for such perfect separation.

**Justification for selecting Random Forest:**

Despite the identical performance on this specific dataset, the **Random Forest Classifier** is selected as the best model for the following reasons:

1.  **Robustness and Generalization**: Random Forests are ensemble methods that build multiple decision trees and merge their predictions. This approach generally leads to better generalization performance and reduced overfitting compared to a single Decision Tree, especially on unseen data or if there were slight variations in the training set.
2.  **Handling Complexity**: For multi-class classification problems like disease prediction, Random Forests often perform more robustly than Naive Bayes, which assumes independence among features. While Naive Bayes also achieved 100% accuracy here, Random Forest's ability to model complex interactions between features makes it a safer choice for real-world application where perfect feature independence might not hold.
3.  **Feature Importance**: Random Forests inherently provide a measure of feature importance, which can be valuable for understanding which symptoms are most indicative of a particular disease. (While not explicitly used in this task, it's a beneficial characteristic of the model).

Therefore, the `RandomForestClassifier` (`rfc`) will be chosen as the `best_model`.

**Reasoning**:
Following the instructions, I need to assign the `rfc` (Random Forest Classifier) object to a new variable named `best_model`.



In [ ]:
best_model = rfc
print("Random Forest Classifier selected as best_model.")

Random Forest Classifier selected as best_model.


## Develop Prediction Function

### Subtask:
Create a Python function that takes a list of symptom names as input, converts these symptoms into a 132-length binary feature vector, and uses the selected best model to predict the disease.


**Reasoning**:
To fulfill the subtask, I need to define a Python function that converts a list of symptoms into a binary feature vector, uses the `best_model` to predict the disease, and then decodes the prediction back to the disease name using the `label_encoder`. This requires `numpy` for array manipulation and access to `X.columns` for mapping symptom names to feature vector indices.



In [ ]:
import numpy as np

def predict_disease(symptoms_list):
    # Create a zero-initialized NumPy array of length 132 (number of features in X)
    feature_vector = np.zeros(len(X.columns))

    # Iterate through the symptoms_list and set corresponding feature to 1
    for symptom in symptoms_list:
        if symptom in X.columns:
            feature_vector[X.columns.get_loc(symptom)] = 1
        else:
            print(f"Warning: Symptom '{symptom}' not found in the feature list.")

    # Reshape the feature vector to be a 2D array for prediction
    feature_vector = feature_vector.reshape(1, -1)

    # Use the best_model to predict the encoded disease label
    predicted_encoded_label = best_model.predict(feature_vector)

    # Convert the predicted encoded label back into the original disease name
    predicted_disease = label_encoder.inverse_transform(predicted_encoded_label)

    return predicted_disease[0]

print("Prediction function `predict_disease` defined.")

Prediction function `predict_disease` defined.


**Reasoning**:
The prediction function `predict_disease` has been successfully defined. To complete the subtask, it's necessary to demonstrate its usage with a sample prediction to ensure it works as expected.



In [ ]:
sample_symptoms = ['itching', 'skin_rash', 'nodal_skin_eruptions', 'dischromic _patches']
predicted_disease = predict_disease(sample_symptoms)
print(f"Predicted Disease for symptoms {sample_symptoms}: {predicted_disease}")

sample_symptoms_2 = ['vomiting', 'abdominal_pain', 'diarrhoea']
predicted_disease_2 = predict_disease(sample_symptoms_2)
print(f"Predicted Disease for symptoms {sample_symptoms_2}: {predicted_disease_2}")

Predicted Disease for symptoms ['itching', 'skin_rash', 'nodal_skin_eruptions', 'dischromic _patches']: Fungal infection
Predicted Disease for symptoms ['vomiting', 'abdominal_pain', 'diarrhoea']: Gastroenteritis


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


**Reasoning**:
The previous code generated a UserWarning because the prediction function's input (a numpy array) lacked feature names, which the Random Forest model was fitted with. To resolve this, I will modify the `predict_disease` function to convert the feature vector into a pandas DataFrame with the correct column names before prediction, ensuring consistency with the training data.



In [ ]:
import numpy as np
import pandas as pd # Import pandas for DataFrame creation

def predict_disease(symptoms_list):
    # Create a zero-initialized NumPy array of length 132 (number of features in X)
    feature_vector = np.zeros(len(X.columns))

    # Iterate through the symptoms_list and set corresponding feature to 1
    for symptom in symptoms_list:
        if symptom in X.columns:
            feature_vector[X.columns.get_loc(symptom)] = 1
        else:
            print(f"Warning: Symptom '{symptom}' not found in the feature list.")

    # Convert the feature vector to a DataFrame with column names for prediction
    # This ensures the model receives input with feature names, suppressing the warning
    input_df = pd.DataFrame([feature_vector], columns=X.columns)

    # Use the best_model to predict the encoded disease label
    predicted_encoded_label = best_model.predict(input_df)

    # Convert the predicted encoded label back into the original disease name
    predicted_disease = label_encoder.inverse_transform(predicted_encoded_label)

    return predicted_disease[0]

print("Prediction function `predict_disease` defined (with feature names handling).")

Prediction function `predict_disease` defined (with feature names handling).


**Reasoning**:
The `predict_disease` function has been updated to handle feature names, addressing the previous UserWarning. I will now re-run the sample predictions to demonstrate that the function works correctly and the warning is suppressed.



In [ ]:
sample_symptoms = ['itching', 'skin_rash', 'nodal_skin_eruptions', 'dischromic _patches']
predicted_disease = predict_disease(sample_symptoms)
print(f"Predicted Disease for symptoms {sample_symptoms}: {predicted_disease}")

sample_symptoms_2 = ['vomiting', 'abdominal_pain', 'diarrhoea']
predicted_disease_2 = predict_disease(sample_symptoms_2)
print(f"Predicted Disease for symptoms {sample_symptoms_2}: {predicted_disease_2}")

Predicted Disease for symptoms ['itching', 'skin_rash', 'nodal_skin_eruptions', 'dischromic _patches']: Fungal infection
Predicted Disease for symptoms ['vomiting', 'abdominal_pain', 'diarrhoea']: Gastroenteritis


## Final Task

### Subtask:
Summarize the complete machine learning pipeline, including data loading, EDA, preprocessing, model training, evaluation, selection, and the final prediction function, highlighting the chosen best model and a sample prediction.


## Summary:

### Data Analysis Key Findings

*   The `Training.csv` dataset, initially comprising 4920 rows and 134 columns, was loaded successfully.
*   Exploratory Data Analysis revealed 41 unique diseases, with a perfectly balanced distribution of 120 instances for each disease, indicating no class imbalance.
*   An irrelevant column, 'Unnamed: 133', containing 4920 missing values (100% of its entries), was identified and removed, resulting in a clean dataset with 133 columns and no missing values.
*   The dataset was preprocessed by separating features (X, shape \$(4920, 132)\$) and the target variable ('prognosis'), which was then numerically encoded into 41 unique labels.
*   The data was split into training (\$(3444, 132)\$) and testing (\$(1476, 132)\$) sets using a 70/30 split with stratification to maintain class distribution.
*   Decision Tree, Random Forest, and Naive Bayes models were trained and evaluated, with all three models achieving a perfect 1.00 accuracy on the test set.
*   The Random Forest Classifier was selected as the best model due to its inherent robustness, better generalization capabilities, and ability to handle complex interactions between features, despite all models showing identical perfect accuracy on this specific dataset.
*   A `predict_disease` function was successfully developed. It takes a list of symptoms, converts them into a properly formatted feature vector (as a Pandas DataFrame with feature names), and uses the selected Random Forest model to predict the disease. Sample predictions for symptoms like `['itching', 'skin_rash']` correctly yielded 'Fungal infection', and `['vomiting', 'abdominal_pain']` predicted 'Gastroenteritis'.

### Insights or Next Steps

*   The 100% accuracy across multiple models suggests that the dataset may be highly separable based on the provided symptoms, or potentially indicative of a very clean and specific dataset structure. Further validation with external, unseen data or a more diverse dataset would be beneficial to confirm generalizability.
*   To make the disease prediction system practical, the `predict_disease` function could be integrated into a user-friendly application or API, allowing easy input of symptoms and immediate disease prediction.
